src/transform/src/dataflow.rs

```diff
@@ -290,41 +292,55 @@ pub mod monotonic {
                 monotonic,
                 ..
             } => {
-                *monotonic = is_monotonic(input, sources);
+                *monotonic = is_monotonic(input, sources, locals);
                 // Reduce is monotonic iff its input is and it is a "distinct",
                 // with no aggregate values; otherwise it may need to retract.
                 *monotonic && aggregates.is_empty()
             }
             MirRelationExpr::Union { base, inputs } => {
-                let mut monotonic = is_monotonic(base, sources);
+                let mut monotonic = is_monotonic(base, sources, locals);
                 for input in inputs.iter_mut() {
-                    let monotonic_i = is_monotonic(input, sources);
+                    let monotonic_i = is_monotonic(input, sources, locals);
                     monotonic = monotonic && monotonic_i;
                 }
                 monotonic
             }
-            MirRelationExpr::ArrangeBy { input, .. } => is_monotonic(input, sources),
+            MirRelationExpr::ArrangeBy { input, .. } | MirRelationExpr::DeclareKeys { .. } => {
+                is_monotonic(input, sources, locals)
+            }
             MirRelationExpr::FlatMap { input, func, .. } => {
-                let is_monotonic = is_monotonic(input, sources);
+                let is_monotonic = is_monotonic(input, sources, locals);
                 is_monotonic && func.preserves_monotonicity()
             }
             MirRelationExpr::Join { inputs, .. } => {
                 // If all inputs to the join are monotonic then so is the join.
                 let mut monotonic = true;
                 for input in inputs.iter_mut() {
-                    let monotonic_i = is_monotonic(input, sources);
+                    let monotonic_i = is_monotonic(input, sources, locals);
                     monotonic = monotonic && monotonic_i;
                 }
                 monotonic
             }
             MirRelationExpr::Constant { rows: Ok(rows), .. } => {
                 rows.iter().all(|(_, diff)| diff > &0)
             }
-            MirRelationExpr::Threshold { input } => is_monotonic(input, sources),
-            // Let and Negate remain
-            _ => {
+            MirRelationExpr::Threshold { input } => is_monotonic(input, sources, locals),
+            MirRelationExpr::Let { id, value, body } => {
+                let prior = locals.remove(id);
```

Comment:
I'm not sure I follow the `prior` logic here. If there is a `prior`, should it be re-installed afterwards? I guess I could see either 1. ignore `prior` or 2. notice `prior` and put things back afterwards, but this seems to do neither (unless I misunderstand, ofc).

Reply:
Thanks for catching the mistake! The reinstallation of prior has been fixed.